data/metric/metrics.go

```diff
@@ -1,203 +1,110 @@
-		}
-		for _, attr := range ms.nsAttributes {
-			if name == attr.Key && strVal == attr.Value {
-				return ErrOverrideSetAttrs
-			}
-		}
-	default:
-		return fmt.Errorf("unknown source type for key %s", name)
-	}
-
-	ms.Metrics[name] = newValue
-
-	return
+type metricBase struct {
+	Timestamp  int64      `json:"timestamp"`
+	Name       string     `json:"name"`
+	Type       string     `json:"type"`
+	Dimensions Dimensions `json:"attributes"`
 }
 
-func (ms *Set) setSetAttribute(name string, value string) {
-	ms.Metrics[name] = value
+// Gauge is a metric of type gauge
+type Gauge struct {
+	metricBase
+	Value float64 `json:"value"`
 }
 
-func castToFloat(value interface{}) (float64, error) {
-	if b, ok := value.(bool); ok {
-		if b {
-			return 1, nil
-		}
-		return 0, nil
-	}
-
-	parsedValue, err := strconv.ParseFloat(fmt.Sprintf("%v", value), 64)
-	if err != nil {
-		return 0, err
-	}
-
-	if isNaNOrInf(parsedValue) {
-		return 0, ErrNonNumeric
-	}
-
-	return parsedValue, nil
+// Count is a metric of type count
+type Count struct {
+	metricBase
+	Interval int64 `json:"interval.ms"`
+	Count    int64 `json:"count"`
 }
 
-func isNaNOrInf(f float64) bool {
-	return math.IsNaN(f) || math.IsInf(f, 0) || math.IsInf(f, -1)
+// Summary is a metric of type summary.
+type Summary struct {
+	metricBase
+	Interval int64   `json:"interval.ms"`
+	Count    int64   `json:"count"`
+	Average  float64 `json:"average"`
+	Sum      float64 `json:"sum"`
+	Min      float64 `json:"min"`
+	Max      float64 `json:"max"`
 }
 
-func (ms *Set) elapsedDifference(name string, absolute interface{}, sourceType SourceType) (elapsed float64, err error) {
-	if ms.storer == nil {
-		err = ErrNoStoreToCalcDiff
-		return
-	}
-
-	newValue, err := castToFloat(absolute)
-	if err != nil {
-		err = ErrNonNumeric
-		return
-	}
-
-	// Fetch last value & time
-	var oldValue float64
-	oldTime, err := ms.storer.Get(ms.namespace(name), &oldValue)
-	if err != nil && err != persist.ErrNotFound {
-		return
-	}
-
-	// Store new value & time (no IO flush until Save)
-	newTime := ms.storer.Set(ms.namespace(name), newValue)
-
-	// First value
-	if err == persist.ErrNotFound {
-		return 0, nil
-	}
-
-	// Time constraints
-	duration := newTime - oldTime
-	if duration == 0 {
-		err = ErrTooCloseSamples
-		return
-	}
-
-	elapsed = newValue - oldValue
-
-	if elapsed < 0 && sourceType.IsPositive() {
-		err = ErrNegativeDiff
-		return
+// NewGauge creates a new metric of type Gauge
+func NewGauge(timestamp time.Time, name string, value float64) Metric {
+	return &Gauge{
+		metricBase: metricBase{
+			Timestamp:  timestamp.Unix(),
+			Name:       name,
+			Type:       SourcesTypeToName[GAUGE],
+			Dimensions: Dimensions{},
+		},
+		Value: value,
 	}
+}
 
-	if sourceType == RATE {
-		elapsed = elapsed / float64(duration)
+// NewCount creates a new metric of type Count
+func NewCount(timestamp time.Time, interval time.Duration, name string, count int64) Metric {
```

Comment:
count can only be positive. Should we use a uint64 or return an error if the value is negative?

Reply:
better a uint64. At first I had uint64 but then changed for some reason :)